eclair-core/src/main/scala/fr/acinq/eclair/io/MessageRelay.scala

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2021 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.io
+
+import akka.actor.ActorRef
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import fr.acinq.bitcoin.Crypto.PublicKey
+import fr.acinq.eclair.wire.protocol.OnionMessage
+
+object MessageRelay {
+  def apply(switchboard: ActorRef, nextNodeId: PublicKey, msg: OnionMessage): Behavior[PeerConnection.ConnectionResult] = {
```

Comment:
Done.
Although I strongly dislike this "best practice", it creates two big regressions:
- using `MessageRelay` is now more complicated, you need to call `MessageRelay()`, spawn a new actor and send it a command instead of just calling a single function
- the actor now has two states that accept different types of messages so we lose some of the benefits from using typed actors.

Reply:
> using MessageRelay is now more complicated, you need to call MessageRelay(), spawn a new actor and send it a command instead of just calling a single function

Well, following the same reasoning, you could say that using multiple actors instead of a single actor that does everything is more complicated (from the caller's point of view), but I think you would agree that it's better engineering?

You should think about actor spawning as a constructor. You usually don't put your business logic inside a constructor, do you? You put it in methods on your object, right? Because that is separation of concern / modularity, which lets your component be extended more easily for unknown future use-cases, and be more easily unit tested.

Reply:
It is normal to split the logic into multiple functions / actors but if the caller needs to know about it that's a design defect.

If the objects needs to be initialized, then I think this initialization logic should go into the constructor, not in a separate `init` function, it's the same thing as `var x; x = "toto";` vs `var x = "toto;"`, I think you agree that the latter is better and that's why modern languages don't allow the former.

Reply:
I completely agree with both of your points, but I don't see how they apply to the current scenario though...?

> if the caller needs to know about it that's a design defect.

I agree with this general statement, but in our case the caller wants to create an actor and have this new actor relay a message. So all of this is part of what the caller expects, it doesn't expose to the caller something he shouldn't know about. The caller creates an actor, then asks it to relay a message: that looks like a perfectly normal calling pattern to me.

Maybe we're talking past each other because you're thinking about it from a functional programming point of view, as just a unique function call (and want to abstract away the fact that it has to create an actor to perform the task)? But I think actor programming is more of an object-oriented paradigm, where you create actors and then ask them to perform actions. The way I view it it feels like a better separation of concern, but maybe that's just me.

> I think this initialization logic should go into the constructor

I agree as well, but sending an outgoing message is not part of an initialization logic, it's an external side-effect. I think that probably pinpoints the core of our misunderstanding: if sending this outgoing message feels part of the internal initialization logic to you, your points make sense.

My point is then simply that calling another actor to trigger business logic doesn't feel part of the initialization logic of the actor to me.

I don't think it's worth arguing too much about this now, especially since the actor is currently so simple, but hopefully once we extend this actor with more initialization logic and more commands to handle, it will make sense why this separation of concern is better (e.g. when it may have to handle various first actions).